pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -0,0 +1,1150 @@
+    Ok(local_layer_mtime)
+}
+
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum EvictionError {
+    #[error("layer was already evicted")]
+    NotFound,
+
+    /// Evictions must always lose to downloads in races, and this time it happened.
+    #[error("layer was downloaded instead")]
+    Downloaded,
+}
+
+/// Error internal to the [`LayerInner::get_or_maybe_download`]
+#[derive(Debug, thiserror::Error)]
+enum DownloadError {
+    #[error("timeline has already shutdown")]
+    TimelineShutdown,
+    #[error("no remote storage configured")]
+    NoRemoteStorage,
+    #[error("context denies downloading")]
+    ContextAndConfigReallyDeniesDownloads,
+    #[error("downloading is really required but not allowed by this method")]
+    DownloadRequired,
+    /// Why no error here? Because it will be reported by page_service. We should had also done
+    /// retries already.
+    #[error("downloading evicted layer file failed")]
+    DownloadFailed,
+    #[error("downloading failed, possibly for shutdown")]
+    DownloadCancelled,
+    #[error("pre-condition: stat before download failed")]
+    PreStatFailed(#[source] std::io::Error),
+    #[error("post-condition: stat after download failed")]
+    PostStatFailed(#[source] std::io::Error),
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum NeedsDownload {
+    NotFound,
+    NotFile,
+    WrongSize { actual: u64, expected: u64 },
+}
+
+impl std::fmt::Display for NeedsDownload {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            NeedsDownload::NotFound => write!(f, "file was not found"),
+            NeedsDownload::NotFile => write!(f, "path is not a file"),
+            NeedsDownload::WrongSize { actual, expected } => {
+                write!(f, "file size mismatch {actual} vs. {expected}")
+            }
+        }
+    }
+}
+
+/// Holds both Arc requriring that both components stay resident while holding this alive and no evictions
+/// nor garbage collection happens.
+#[derive(Clone)]
+pub(crate) struct ResidentLayer {
+    owner: Layer,
+    downloaded: Arc<DownloadedLayer>,
+}
+
+impl std::fmt::Display for ResidentLayer {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.owner)
+    }
+}
+
+impl std::fmt::Debug for ResidentLayer {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.owner)
+    }
+}
+
+impl ResidentLayer {
+    /// Release the eviction guard, making this back into a plain [`Layer`].
+    pub(crate) fn drop_eviction_guard(self) -> Layer {
+        self.into()
+    }
+
+    /// Loads all keys stored in the layer. Returns key, lsn and value size.
+    pub(crate) async fn load_keys(
+        &self,
+        ctx: &RequestContext,
+    ) -> anyhow::Result<Vec<DeltaEntry<'_>>> {
+        use LayerKind::*;
+
+        let owner = &self.owner.0;
+
+        match self.downloaded.get(owner).await? {
```

Comment:
Code order: when reading file top to bottom, downloaded.get() impl hasn't been shown yet.

Wondering why DownloadedLayer doesn't come before ResidentLayer in this file.

It's been a pleasure to read otherwise.

Reply:
No reason, should reorder, it took me a while to understand how simple the ResidentLayer is, so that's why.

Reply:
Reordered now, ResidentLayer is the last in file.